src/amo/components/AddonRecommendations/index.js

```diff
@@ -0,0 +1,169 @@
+/* @flow */
+import makeClassName from 'classnames';
+import { oneLine } from 'common-tags';
+import invariant from 'invariant';
+import * as React from 'react';
+import defaultCookie from 'react-cookie';
+import { compose } from 'redux';
+import { connect } from 'react-redux';
+
+import AddonsCard from 'amo/components/AddonsCard';
+import {
+  fetchRecommendations,
+  getRecommendationsByGuid,
+} from 'amo/reducers/recommendations';
+import { withErrorHandler } from 'core/errorHandler';
+import translate from 'core/i18n/translate';
+import defaultTracking from 'core/tracking';
+import type {
+  Recommendations,
+  RecommendationsState,
+} from 'amo/reducers/recommendations';
+import type { ErrorHandlerType } from 'core/errorHandler';
+import type { I18nType } from 'core/types/i18n';
+import type { AddonType } from 'core/types/addons';
+import type { DispatchFunc } from 'core/types/redux';
+
+export const TAAR_IMPRESSION_CATEGORY = 'AMO Addon / Recommendations Shown';
+export const TAAR_COHORT_COOKIE_NAME = 'TAARCohort';
+export const TAAR_COHORT_INCLUDED: 'TAAR_COHORT_INCLUDED'
+  = 'TAAR_COHORT_INCLUDED';
+export const TAAR_COHORT_EXCLUDED: 'TAAR_COHORT_EXCLUDED'
+  = 'TAAR_COHORT_EXCLUDED';
+
+export type CohortName = typeof TAAR_COHORT_INCLUDED | typeof TAAR_COHORT_EXCLUDED;
+
+type Props = {|
+  cookie: typeof defaultCookie,
+  randomizer: () => number,
+  tracking: typeof defaultTracking,
+  addon: AddonType | null,
+  className?: string,
+  dispatch: DispatchFunc,
+  errorHandler: ErrorHandlerType,
+  i18n: I18nType,
+  recommendations: Recommendations | null,
+|};
+
+
+export class AddonRecommendationsBase extends React.Component<Props> {
+  static defaultProps = {
+    cookie: defaultCookie,
+    randomizer: Math.random,
+    tracking: defaultTracking,
+    recommendations: null,
+  };
+
+  componentWillMount() {
```

Comment:
I don't think this will be working as intended, for instance, I cannot get any recommendations from the server. It works well in client side mode but not in server side mode:

![screen shot 2018-05-07 at 10 49 55](https://user-images.githubusercontent.com/217628/39694853-ef9546e2-51e8-11e8-81e1-33a83dcb4f64.png)

I think you can reproduce this issue by navigating to an addon detail page and hit "reload" to fully reload the page. It does not work because:

1. the server sends the request to fetch the recommandations
2. the state sets a new `recommendation` entry to the recommendations state (with `loading: true`)
3. the server `END`s the sagas but the recommendations saga was not finished
4. the client loads the preloaded state (from the server) with the `recommendation` entry to the recommendations state (with `loading: true`), 💥 deadlock

I see that we explicitly by-pass the server side rendering in `AddAddonToCollection`: https://github.com/mozilla/addons-frontend/blob/2111ff06c6a9499d39a7d1db29d828285b6986ee/src/amo/components/AddAddonToCollection/index.js#L67-L72

Reply:
Using the `componentDidMount()` method instead of `componentWillMount()` would work here.

Reply:
In addition: the screenshot above was taken to show a UI issue (the card height is not big enough)

Reply:
Can we move everything in `componentWillMount` to `componentDidMount`, including the cookie setting code, or do we just want to move the call to `this.dispatchFetchRecommendations` into `componentDidMount`?

Also, why is this an issue with this particular component? It seems like a common pattern we have to call a fetch action in `componentWillMount` and it isn't causing a problem in other components where we do it.

Reply:
> Can we move everything in componentWillMount to componentDidMount, including the cookie setting code, or do we just want to move the call to this.dispatchFetchRecommendations into componentDidMount?

I guess, yes.

> Also, why is this an issue with this particular component? It seems like a common pattern we have to call a fetch action in componentWillMount and it isn't causing a problem in other components where we do it.

It depends on the components I think. Because of the way we fetch the data and stops the sagas on the server, we have to be very careful. For some components, we by-pass data fetching on the server, for some others, the parent component gets the data and we also have the case where there is no `loading` indicator, which basically allows the client to fetch the data again.

Reply:
Thanks. I tried to fix the issue with the size of the card when it's loading, but I couldn't get it to work. Increasing the padding at the bottom increased the distance from the bottom of the box to the footer, but didn't increase the size of the box (because of the white background, I think). This may require some more time and/or better css chops, so I think I want to defer that to a follow-up.

Reply:
Sure, I was thinking maybe that's because we have 4 addons here. Can we have 3 "loading" addons instead? Not sure if that would change anything.

Reply:
3 doesn't make it any better. :(

Reply:
D'oh! Could you please open an issue so that we don't forget?